js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,479 @@
+		}
+		opts = append(opts, grpc.WithContextDialer(dialer))
+
+		ctx, cancel := context.WithTimeout(*ctxPtr, timeout)
+		defer cancel()
+
+		var err error
+		c.conn, err = grpc.DialContext(ctx, addr, opts...)
+		if err != nil {
+			errc <- err
+
+			return
+		}
+		close(errc)
+	}()
+
+	if err := <-errc; err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// InvokeRPC creates and calls a unary RPC by fully qualified method name
+//nolint: funlen,gocognit
+func (c *Client) InvokeRPC(ctxPtr *context.Context,
+	method string, req goja.Value, params map[string]interface{}) (*Response, error) {
+	ctx := *ctxPtr
+	rt := common.GetRuntime(ctx)
+	state := lib.GetState(ctx)
+	if state == nil {
+		return nil, errInvokeRPCInInitContext
+	}
+
+	if c.conn == nil {
+		return nil, errors.New("no gRPC connection, you must call connect first")
+	}
+
+	method = strings.TrimPrefix(method, "/")
+	md := c.mds[method]
+
+	if md == nil {
+		return nil, fmt.Errorf("method %q not found in file descriptors", method)
+	}
+
+	c.tags = state.CloneTags()
+	timeout := 60 * time.Second
+
+	ctx = metadata.NewOutgoingContext(ctx, metadata.New(nil))
+	for k, v := range params {
+		switch k {
+		case "headers":
+			rawHeaders, ok := v.(map[string]interface{})
+			if !ok {
+				continue
+			}
+			for hk, kv := range rawHeaders {
+				// TODO(rogchap): Should we manage a string slice?
+				strVal, ok := kv.(string)
+				if !ok {
+					continue
+				}
+				ctx = metadata.AppendToOutgoingContext(ctx, hk, strVal)
+			}
+		case "tags":
+			rawTags, ok := v.(map[string]interface{})
+			if !ok {
+				continue
+			}
+			for tk, tv := range rawTags {
+				strVal, ok := tv.(string)
+				if !ok {
+					continue
+				}
+				c.tags[tk] = strVal
+			}
+		case "timeout":
+			if t, ok := v.(float64); ok && t > 0.0 {
+				timeout = time.Duration(t) * time.Millisecond
+			}
+		}
+	}
+	if state.Options.SystemTags.Has(stats.TagURL) {
+		c.tags["url"] = fmt.Sprintf("%s/%s", c.conn.Target(), method)
+	}
+
+	parts := strings.Split(method, "/")
+	if state.Options.SystemTags.Has(stats.TagService) {
+		c.tags["service"] = parts[0]
+	}
+	if state.Options.SystemTags.Has(stats.TagMethod) {
+		c.tags["method"] = parts[1]
+	}
+
+	if state.Options.SystemTags.Has(stats.TagRPCType) {
+		// (rogchap) This method only supports unary RPCs
+		// if this is refactored to support streaming then this should
+		// be updated to be based on the method descriptor (IsClientStreaming/IsServerStreaming)
+		c.tags["rpc_type"] = "unary"
+	}
+
+	// Only set the name system tag if the user didn't explicitly set it beforehand
+	if _, ok := c.tags["name"]; !ok && state.Options.SystemTags.Has(stats.TagName) {
+		c.tags["name"] = method
+	}
+
+	reqdm := dynamic.NewMessage(md.GetInputType())
+	b, _ := req.ToObject(rt).MarshalJSON()
+	_ = reqdm.UnmarshalJSON(b)
+
+	if rpsLimit := state.RPSLimit; rpsLimit != nil {
+		if err := rpsLimit.Wait(ctx); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Fair enough; but as of today, this is the approach for the HTTP requests and we need the same option to limit rps for our load tests on gRPC.

Reply:
Hi, I asked @rogchap (we work for the same company) for the `--rps` option for a particular use case where I wanted a constant number of VUs and an overall rate limit. I couldn't see how to do that with arrival-rate executors because they appear to modulate the number of VUs according to the desired request rate.

The reason I required a constant number of VUs was to provide a full distribution of data sharded uniquely across each VU.

I understand, from a chat and emails with @robingustafsson that this may be addressed in the mid-January release. #532 and #1539. 

At the time, I didn't realise `--rps` is/should be deprecated, and am open to other suggestions. Another option I considered was just adding a sleep function in the VUs, but felt like that was fighting the tool, so `--rps` seemed to be the natural option.

Reply:
> The reason I required a constant number of VUs was to provide a full distribution of data sharded uniquely across each VU.
> 
> I understand, from a chat and emails with @robingustafsson that this may be addressed in the mid-January release. #532 and #1539.

@shric, yeah, the current plan is to work on those after k6 v0.29.0, though it's a bit early to tell if we can manage to land them in v0.30.0 yet. And I'm still not sure I understand your use case very well, so can you please explain it in detail in https://github.com/loadimpact/k6/issues/1539, it would help us a lot when designing the feature.

Specifically, I am unsure where these 2 statements are in conflict and why that necessitates `--rps`:

> I couldn't see how to do that with arrival-rate executors because they appear to modulate the number of VUs according to the desired request rate.
> The reason I required a constant number of VUs was to provide a full distribution of data sharded uniquely across each VU.

Try running the following k6 script:
```js
import http from 'k6/http';

export let options = {
    "scenarios": {
        "arrival": {
            "executor": "ramping-arrival-rate",
            "preAllocatedVUs": 5,
            "stages": [
                { "target": 2, "duration": "10s" },
                { "target": 10, "duration": "10s" },
            ],
        },
    }
};

export default function () {
    console.log(__VU);
    http.get("https://test.k6.io")
}
```

See how all 5 VUs are used in a roughly round-robin fashion, despite the test run actually requiring only a single VU executing an iteration at any point in time. This happens because the initialized VU pool is implemented as a buffered Go channel ([global pool](https://github.com/loadimpact/k6/blob/62b564f63478100897093720895feedc15078a50/lib/execution.go#L154-L177), [`ramping-arrival-rate` pool](https://github.com/loadimpact/k6/blob/62b564f63478100897093720895feedc15078a50/lib/executor/ramping_arrival_rate.go#L326)), so it's essentially a FIFO queue. When the `ramping-arrival-rate` takes a VU from the "buffer pool" to run an iteration on it, it's going to take it from the start of the channel, and when the iteration is done, it's going to put the VU back at the end of the channel. So, not quite round-robin (VUs might be returned out of order when their iterations take different times to execute), but k6 should cycle between all initialized VUs fairly evenly during the test run, even if only a few of them are required at any given point in time. 

I don't fully grok your use case, but does this resolve it? Each VU should be able to have its turn making requests, so all of the data will be used eventually. Or am I missing something?

> At the time, I didn't realise `--rps` is/should be deprecated

`--rps` for HTTP calls is probably never going to be completely deprecated, since a lot of old pre-arrival-rate scripts use it... But it's a very confusing option that doesn't work well with distributed execution, so it's likely we'll heavily encourage users to avoid using it and rely on `arrival-rate` executors instead. They probably won't address 100% of the use cases, but the remainder should probably be addressed via the new HTTP API we plan to make... eventually... So, for the foreseeable future, `--rps` for HTTP won't be deprecated, just discouraged.

So, if we can avoid adding `--rps` in a the brand new gRPC API from the start, I'd prefer we do so.

Reply:
Ah yes, that explains it and your example works completely fine. I had falsely assumed that during a test run something like: `foo [=>------] 01/10 VUs  0m46.6s/5m0s  3 iters/s` meant that *only* __VU 1 was running, rather than meaning 1 VU is running concurrently but they are round-robin, which is great. It would be good to document this more clearly in https://k6.io/docs/using-k6/scenarios or the relevant executors' docs.

Agreed, my use case doesn't need `--rps`, the `ramping-arrival-rate`  executor is perfect already. However, a user might get surprised (as I was before Roger added --rps) if they use `--rps` and find it has no effect on grpc but only http, so that should be clearly documented in the usage as well?

Reply:
I added https://github.com/loadimpact/k6-docs/issues/136#issue-723076528 to not forget to fix the docs